### PR TITLE
Update `BlockV2` to hold `TransactionHash`es

### PR DESCRIPTION
**Pull request by Fraser999**
This PR updates the `BlockV2` types to handle `Transaction`s rather than just `Deploy`s.

As discussed in a recent standup, the `UserlandTransactionV1::Noop` variant was removed.

Closes [roadmap#183](https://github.com/casper-network/roadmap/issues/183).

### CODE REVIEW

File: types/src/transaction.rs [hunk 1]
```diff
@@ -109,6 +110,24 @@ impl Transaction {
         }
     }
 
+    /// Returns the computed approvals hash identifying this transaction's approvals.
+    pub fn compute_approvals_hash(&self) -> TransactionApprovalsHash {
```
Comment (alsrdn) [inline types/src/transaction.rs]: Any reason not to return a `Result<TransactionApprovalsHash>` here?

File: node/src/components/block_synchronizer/block_acquisition.rs [hunk 2]
```diff
@@ -584,7 +584,16 @@ impl BlockAcquisitionState {
...
                     header.block_hash()
                 );
-                let deploy_hashes = block.deploy_and_transfer_hashes().copied().collect();
+                let deploy_hashes = match &block {
+                    Block::V1(v1) => v1.deploy_and_transfer_hashes().copied().collect(),
+                    Block::V2(v2) => v2
+                        .all_transactions()
+                        .filter_map(|txn_hash| match txn_hash {
+                            TransactionHash::Deploy(deploy_hash) => Some(*deploy_hash),
+                            TransactionHash::V1(_) => None,
```
Comment (alsrdn) [inline node/src/components/block_synchronizer/block_acquisition.rs:593]: Wouldn't we need a separate `transaction_acquisition` here for transactions identified by `TransactionHash`?

File: node/src/components/block_synchronizer/block_acquisition.rs [hunk 3]
```diff
@@ -1032,7 +1041,16 @@ impl BlockAcquisitionState {
...
                     block.hash()
                 );
-                let deploy_hashes = block.deploy_and_transfer_hashes().copied().collect();
+                let deploy_hashes = match block.as_ref() {
+                    Block::V1(v1) => v1.deploy_and_transfer_hashes().copied().collect(),
+                    Block::V2(v2) => v2
+                        .all_transactions()
+                        .filter_map(|txn_hash| match txn_hash {
+                            TransactionHash::Deploy(deploy_hash) => Some(*deploy_hash),
+                            TransactionHash::V1(_) => None,
```
Comment (alsrdn) [inline node/src/components/block_synchronizer/block_acquisition.rs:1050]: Same question as above for `register_block`.

File: node/src/components/block_validator.rs [hunk 4]
```diff
@@ -188,12 +202,12 @@ where
...
                 debug!(?block, "validating proposed block");
                 if block.deploy_hashes().count()
-                    > self.chainspec.transaction_config.block_max_deploy_count as usize
+                    > self.chainspec.transaction_config.block_max_standard_count as usize
                 {
                     return responder.respond(false).ignore();
                 }
                 if block.transfer_hashes().count()
-                    > self.chainspec.transaction_config.block_max_native_count as usize
+                    > self.chainspec.transaction_config.block_max_transfer_count as usize
```
Comment (alsrdn) [inline node/src/components/block_validator.rs:210]: We should check for `block_max_staking_count` and `block_max_install_upgrade_count` here as well.

File: node/src/components/contract_runtime/operations.rs [hunk 5]
```diff
@@ -131,8 +137,13 @@ pub fn execute_finalized_block(
...
     }
 
-    // WARNING: Do not change the order of `deploys` as it will result in a different root hash.
-    for deploy in executable_block.deploys {
+    // WARNING: Do not change the order of `transactions` as it will result in a different root
+    // hash.
+    for txn in executable_block.transactions {
+        let deploy = match txn {
+            Transaction::Deploy(deploy) => deploy,
+            Transaction::V1(_) => continue,
```
Comment (alsrdn) [inline node/src/components/contract_runtime/operations.rs:145]: Why are we not executing V1 transactions here?

File: node/src/reactor/main_reactor/control.rs [hunk 6]
```diff
@@ -375,7 +375,10 @@ impl MainReactor {
         // sufficient finality signatures have been collected.
```
Comment (alsrdn) [inline node/src/reactor/main_reactor/control.rs:375]: nit for comment: this genesis block has no *transactions*

**Review by alsrdn — COMMENTED** [review]
Did a first pass of this PR. Posting some questions.

File: types/src/transaction.rs (hunk 1 above)
Comment (Fraser999) [reply types/src/transaction.rs]: Nope - good point.  Fixed in [a118d05](https://github.com/casper-network/casper-node/pull/4392/commits/a118d0591f945a731adfc0521ebf3a8ffe8aea2b).

File: node/src/reactor/main_reactor/control.rs (hunk 6 above)
Comment (Fraser999) [reply node/src/reactor/main_reactor/control.rs:375]: Fixed in [a118d05](https://github.com/casper-network/casper-node/pull/4392/commits/a118d0591f945a731adfc0521ebf3a8ffe8aea2b).

File: node/src/components/block_synchronizer/block_acquisition.rs (hunk 2 above)
Comment (Fraser999) [reply node/src/components/block_synchronizer/block_acquisition.rs:593]: There are still a few tickets outstanding to complete this epic.  These include updating `DeployBuffer`, `BlockSynchronizer`, `ContractRuntime` and `BlockValidator`.  Until those tickets are started, I've been keeping the logic in place in those components to support executing `Deploy`s only.

File: node/src/components/block_synchronizer/block_acquisition.rs (hunk 3 above)
Comment (Fraser999) [reply node/src/components/block_synchronizer/block_acquisition.rs:1050]: As per comment above.

File: node/src/components/block_validator.rs (hunk 4 above)
Comment (Fraser999) [reply node/src/components/block_validator.rs:210]: As per comment above.

File: node/src/components/contract_runtime/operations.rs (hunk 5 above)
Comment (Fraser999) [reply node/src/components/contract_runtime/operations.rs:145]: As per comment above.

**Review by alsrdn — APPROVED** [review]

**Comment by Fraser999** [comment]
bors merge

File: node/src/components/block_synchronizer/tests.rs [hunk 7]
```diff
@@ -1581,6 +1581,18 @@ async fn fwd_having_block_body_for_block_with_deploys_requires_approvals_hashes(
...
     for event in events {
+        if !matches!(
+            event,
+            MockReactorEvent::ApprovalsHashesFetcherRequest(FetcherRequest {
+                id,
+                peer,
+                ..
+            }) if peers.contains(&peer) && id == *block.hash()
+        ) {
+            println!("peers: {:?}", peers);
```
Comment (rafal-ch) [inline node/src/components/block_synchronizer/tests.rs:1592]: Do you want to keep these prints or is this a temporary code?

**Pull request merged**